feat(vite): Add ability to switch Webpack for Vite on the web side

dac09 opened the pull request:
### Experimental Vite Support 🎤


This PR does the following:

- [x] changes within the framework to make it Vite compatible (there's quite a lot here.... see full breakdown: https://www.notion.so/redwoodjs/Vite-RW-95f38a4e57ab4391b172b9d771b75104) 
- [x] adds setup command to opt-in to using Vite. This will modify your project to make it Vite compatible
- [x] updates CLI to look for the bundler flag, and switch commands when running dev
- [x] find a way to make storybook run - currently failing because of the `await import` in DevFatalErrorPage
		Added experiments.topLevelAwait to storybook webpack config! Storybook is still running webpack

**Importantly 👉** Everything should be backwards compatible (so no breaking changes for anyone remaining on webpack)

---

### Outstanding tasks

- [x] Use config hook and setup the preset as a Vite plugin
- [x] Can we "inject" the entry point automatically instead of modifying index.html
- [x] Redirect to root doesn't seem to be working (i.e. refresh the page after navigating, you get a blank page)
- [x] Swap build command to vite
- [x] Fix prerender failures caused by top level await in FatalErrorPage

Tobbe commented:
<img src="https://media1.giphy.com/media/xT9IgfNPRprdidd3Bm/giphy.gif"/>

Tobbe commented on packages/auth-providers-web/ambient.d.ts:
```diff
@@ -0,0 +1,53 @@
+  var RWJS_ENV: {
+    RWJS_API_GRAPHQL_URL: string
+    /** URL or absolute path to the DbAuth serverless function */
+    RWJS_API_DBAUTH_URL: string
```
This shouldn't be here. Created this PR to clean it up https://github.com/redwoodjs/redwood/pull/6968

dac09 replied:
yeah, I'll take it out too! Do we need this ambient file in that case?

Tobbe replied:
I don't think so. But things also worked previously when we didn't have it, even though maybe we should have. Does Vite change something that makes the ambient files more important?

Tobbe commented on packages/cli-helpers/src/lib/installHelpers.ts:
```diff
@@ -0,0 +1,30 @@
+import execa from 'execa'
+
+export const addWebPackages = (webPackages: string[], rwVersion: string) => ({
```
Should this be part of this PR? I like the refactor, just doesn't seem to have anything to do with Vite

dac09 replied:
I need the addWebPackages function in the setup vite command.

Tobbe replied:
Cool. I always just thought of it as an auth specific helper. But of course other commands could use it too! Should we make an auth-cli-helpers package with things only used by the auth setup commands? I guess that kind of ties into the @redwoodjs/paths discussions we've been having. So you don't have to answer. Just something to keep in mind for when we do have the rw/paths conversation next time.

Also don't have to make any code changes here. As I said - I like the refactor 🙂 Just missed that you used it for the vite setup command.

Tobbe commented on packages/auth-providers-web/ambient.d.ts:
```diff
@@ -0,0 +1,53 @@
+import type { HelmetServerState } from 'react-helmet-async'
+
+declare global {
+  var bazinga: boolean
```
```suggestion
```

Tobbe commented on packages/internal/src/build/babel/web.ts:
```diff
@@ -111,7 +113,7 @@ export const getWebSideOverrides = (
     // the `./web/src/Routes.[ts|jsx]` file.
     {
-      test: /Routes.(js|tsx)$/,
+      test: /Routes.(js|tsx|jsx)$/,
```
Do we really support .jsx files?

dac09 replied:
Yes we do - read this for context: https://github.com/vitejs/vite/discussions/3448

I have a plugin temporarily so that you don't need to switch to .jsx, but once Vite is stable, I think we should move everything to .jsx as per the thread.

Tobbe replied:
I knew I remembered something about a js vs jsx conversation! https://github.com/redwoodjs/redwood/issues/734

Tobbe replied:
So we do support .jsx, we just didn't want to make it the default 🙂  (That was a long time ago though, so people might have changed their name)

jtoar replied:
> ...once Vite is stable, I think we should move everything to .jsx as per the thread

We've discussed this once before and a lot of the team was against this. So queue it up for discussion again when you're ready

Tobbe commented on packages/web/src/components/DevFatalErrorPage.tsx:
```diff
@@ -7,7 +7,9 @@ import { useState } from 'react'
 // RWJS_SRC_ROOT is defined and defaulted in webpack to the base path
-const srcRoot = process.env.RWJS_SRC_ROOT || ''
+
+// @MARK error boundary not working in Vite
```
This comment broke the connection between the RWJS_SRC_ROOT comment and the usage of RWJS_SRC_ROOT. Is it possible to move the @MARK somewhere else?

dac09 replied:
Yeah I'll remove this, got it working.

dac09 commented on tasks/framework-tools/lib/framework.mjs:
```diff
@@ -120,6 +120,9 @@ export function frameworkPackagesBins(packages = frameworkPkgJsonFiles()) {
       continue
     }
+
+    // @TODO @MARK this interferes with having a single bin file
```
@jtoar wanted to chat to you about this and the bin aliasing we do!

dac09 commented:
@KrisCoulson ready to test! How to try on your project before the PR is merged:


```
# Copy latest and greatest code from this branch
yarn rwfw project:sync 

# Setup vite in your project
yarn rw setup vite --no-add-package

# Inside redwood/packages/vite - in the framework.
# this will publish the new redwoodjs/vite package locally
yarn build && yarn dlx yalc publish

# Then inside your project
cd web && yarn dlx yalc add @redwoodjs/vite
yarn
```

jtoar commented on .eslintrc.js:
```diff
@@ -59,6 +59,7 @@ module.exports = {
         'packages/forms/src/**',
         'packages/prerender/src/browserUtils/**',
         'packages/router/src/**',
+        'packages/auth-providers-web/src/**',
```
This one's gone now:

```suggestion
```

jtoar commented on packages/auth-providers/dbAuth/web/src/__tests__/dbAuth.test.ts:
```diff
@@ -4,8 +4,10 @@ import { CurrentUser } from '@redwoodjs/auth'
-process.env.RWJS_API_URL = '/.redwood/functions'
-process.env.RWJS_API_GRAPHQL_URL = '/.redwood/functions/graphql'
+// @NOTE there are TS errors assigning these, because there's no ambient declarations
+// But because this is going to be moved as part of decoupling, ignoring for now
```
> // But because this is going to be moved as part of decoupling, ignoring for now

Not decoupling auth right? Some other decoupling? Just clarifying

dac09 replied:
Hey I've fixed this... I misread the path, so thought it wouldn't be needed!

jtoar commented on packages/auth-providers/dbAuth/web/src/dbAuth.ts:
```diff
@@ -49,7 +53,7 @@ export function createDbAuthClient({
 
   const getApiDbAuthUrl = () => {
-    return dbAuthUrl || `${process.env.RWJS_API_URL}/auth`
+    return dbAuthUrl || `${RWJS_API_URL}/auth`
```
Why doesn't this one need `globalThis`? Not saying it should or shouldn't, I just don't understand the nuance

dac09 replied:
Thought I replied to this! Not adding the `globalThis` is the only way I can ensure it works with both Webpack and Vite.

I think there's subtle differences in the way these values are defined between the two bundlers

jtoar commented on packages/web/src/components/DevFatalErrorPage.tsx:
```diff
@@ -7,7 +7,7 @@ import { useState } from 'react'
 
 // RWJS_SRC_ROOT is defined and defaulted in webpack to the base path
-const srcRoot = process.env.RWJS_SRC_ROOT || ''
+const srcRoot = RWJS_DEBUG_ENV?.RWJS_SRC_ROOT || ''
```
I don't understand the nuance yet so just going to ask: why no `globalThis`?

dac09 replied:
When you use `globalThis` - it fails to find the variable in either Webpack or Vite.

Directly accessing the variable like this is the only way I could make sure it works for both!

jtoar commented on packages/cli-helpers/src/lib/installHelpers.ts:
```diff
@@ -0,0 +1,24 @@
+  },
+})
+
+export const addApiPackages = (apiPackages: string[]) =>
+  apiPackages.length > 0 && {
+    title: 'Adding required api packages...',
+    task: async () => {
+      await execa('yarn', ['workspace', 'api', 'add', ...apiPackages])
+    },
+  }
```
This one's changed a bit; the deleted code is up to date but not this one

```suggestion
export const addApiPackages = (apiPackages: string[]) => ({
  title: 'Adding required api packages...',
  task: async () => {
    await execa('yarn', ['workspace', 'api', 'add', ...apiPackages])
})
```

jtoar commented on packages/cli/src/commands/generate/dbAuth/__tests__/dbAuth.test.js:
```diff
@@ -1,8 +1,4 @@
-
-jest.mock('fs')
-
-import fs from 'fs'
```
Seems like we need the `fs` stuff back here: https://github.com/redwoodjs/redwood/actions/runs/3692167967/jobs/6250798520

jtoar commented on packages/cli/src/commands/setup/vite/vite.js:
```diff
@@ -0,0 +1,154 @@
+  })
+}
+
+export const handler = async ({ force, verbose, addPackage }) => {
```
The handler and it's dependencies need to be in its own file, `viteHandler.js` (see `builderHandler.js`, `devHandler.js`, etc) and asynchronously imported here; else the upfront require cost is going to slow down other commands

jtoar replied:
We haven't done this with many setup commands before, but I think we should continue the trend

jtoar replied:
Got a PR here to get the other commands on the same page https://github.com/redwoodjs/redwood/pull/7156

dac09 replied:
Yeah happy to change this, no worries! Feel free to push if you like

jtoar commented on packages/vite/package.json:
```diff
@@ -0,0 +1,37 @@
+    "build": "yarn build:js",
+    "build:js": "babel src -d dist --extensions \".js,.ts,.tsx\""
+  },
+  "dependencies": {
+    "@babel/runtime-corejs3": "7.20.6",
+    "@esbuild-plugins/node-globals-polyfill": "0.1.1",
+    "@originjs/vite-plugin-commonjs": "1.0.3",
+    "@redwoodjs/internal": "3.2.0",
+    "@vitejs/plugin-react": "2.2.0",
+    "core-js": "3.26.1",
+    "vite": "3.2.5",
+    "vite-plugin-html": "3.2.0"
+  },
```
`esbuild` is missing given it's used in the `src/index.ts` file (line 8):

```ts
import { transform } from 'esbuild'
```

dac09 replied:
Esbuild is a dependency of vite, should be ok no?

jtoar replied:
It may be ok, but it may also be susceptible to hoisting since the internal package lists esbuild as a dependency. I'm just going off of yarn's rulebook which says "Packages should only ever require what they formally list in their dependencies" https://yarnpkg.com/advanced/rulebook#packages-should-only-ever-require-what-they-formally-list-in-their-dependencies

jtoar commented on packages/vite/src/index.ts:
```diff
@@ -0,0 +1,170 @@
+import { getConfig } from '@redwoodjs/internal/dist/config'
+import { getPaths } from '@redwoodjs/internal/dist/paths'
+
+const redwoodConfig = getConfig()
+const redwoodPaths = getPaths()
```
These functions are executed as soon as this file is imported, which may not be a problem given the changes I've made to the CLI to ensure the `cwd` is set before the CLI imports anything. But just something to note if you think this file could be executed somehow before `cwd` is set

dac09 replied:
Hmmm... I'll move this inside the function!

jtoar reviewed (commented):
Haven't tried it out yet and don't understand Vite at all so just commented on the things I could initially

jtoar commented on packages/cli/src/commands/setup/vite/templates/vite.config.ts.template:
```diff
@@ -0,0 +1,13 @@
+
+import redwood from '@redwoodjs/vite'
+
+/**
+ * https://vitejs.dev/config/
+ * @type {import('vite').UserConfig}
+ */
+const viteConfig: UserConfig = {
+  plugins: [redwood()],
+}
```
This leaves us with a bit of type duplication in the case of ts files; wonder if we could clean it up? I'll see if it's this way with other commands too

dac09 replied:
ts-to-js doesn't remove these type comments, so its the only way to guarantee both. Do you see an error in your editor? Mine is fine with both the comment and type

jtoar replied:
Nah it's not about an error just about duplication; not something we need to clean up now

jtoar commented on packages/cli/src/commands/setup/vite/vite.js:
```diff
@@ -0,0 +1,154 @@
+          if (config.web.bundler !== 'vite') {
+            config.web.bundler = 'vite'
+
+            writeFile(redwoodTomlPath, toml.stringify(config), {
+              overwriteExisting: true, // redwood.toml always exists
+            })
```
Seems like this step ruins the formatting and deletes the comments:

```toml
[web]
title = "Redwood App"
port = 8_910
apiUrl = "/.redwood/functions"
includeEnvironmentVariables = [ ]
bundler = "vite"

[api]
port = 8_911

[browser]
open = true

```

dac09 replied:
Yeah noticed this too, but don't intend on solving right now.

dac09 replied:
Actually fixed it with a simple string replace!

jtoar commented on packages/cli/src/commands/setup/vite/vite.js:
```diff
@@ -0,0 +1,154 @@
+          // Keep it as JSX for now
+          const entryPointFile = path.join(
+            getPaths().web.src,
+            `entry-client.jsx`
```
Not sure about the name here; why not just call it `index.js` like the one we already have? And this file has to be there? Right now we keep `index.js` under wraps, and give users the option to generate it if they want more power. What's the nuance or reasoning for changing?

dac09 replied:
I've intentionally called it entry-client, because for render modes you need a different entry point for server rendering. 

I know I originally hid the index.js, but in retrospect I think it doesn't change as often as I thought it did. We also didn't have codemod capability at the time!

jtoar reviewed (commented):
Another quick round

dac09 commented:
@jtoar ready for merge me thinks - addressed all comments, and fixed the issues @KrisCoulson and I found the other day! Also bumped to Vite 4. Note that I see an issue in storybook that looks like an unrelated issue: https://github.com/storybookjs/storybook/pull/20383

It's still experimental ofcourse, and I will need to add fixes as we find them, but its now in a state where I think I'd like to get into more hands

FYI there's some stuff I've intentionally not done yet, e.g. 

- [ ] Vite dev and vite build should be called programatically
- [ ] Prerendering images causes images to reload 

Next step for me is to write up a guide for people to try over the holidays!

dac09 commented:
@Tobbe would you like to review again/dismiss the old review please? 

Want to get this into main (pushing up with merge conflicts resolved and new fixture shortly)

Tobbe commented:
@dac09 I'll try to give it another look after our stand-up meeting later today

Tobbe reviewed (approved):
I'm happy with your changes/replies to my comments, but please let @jtoar go through his comments too before you merge

jtoar commented:
I think we can see about listing esbuild as a dependency later. I'm inclined to just cause it's yarn's rules, but let's see what happens hoisting-wise once it's merged and the package is published. I'll also separate the handler after merge—I don't want to risk messing it up while it's in a workable state, and I'm only just getting back into things. One thing I'm wondering is if the toml change users have to make should just be a note logged to the console till the toml stuff is figured out. But that's also something we can do after merge.

Great stuff @dac09! I'll merge this so I can publish a canary version of the new package.
